scripts/CreateMirrorHierarchy.pl

```diff
@@ -132,64 +132,137 @@ sub CanextendCheck {
+if( system("vgdisplay '$VGName' > /dev/null 2>&1") ){
+	print "Volume group $VGName does not exist, creating now.\n";
+	
+	# Create the volume group
+	$vgret = system("vgcreate -f $VGName ${opt_l} 2>/dev/null");
+}else{
+	print "Volume group $VGName already exists, skipping create of the volume group.\n";
 }
 
-# Check that the specified devices exists on the target server
-@output = `$LKDIR/bin/lcdremexec -d $targetSys -- "if [ ! -b $opt_r ]; then echo no; else echo yes; fi"`;
-if (($? != 0) || (grep(/^no/, @output))) {
-	print "Specified device $opt_r does not exist on $targetSys\n";
+if ($vgret != 0) {
+	print "Failed to create Volume Group $VGName for ${opt_l} on the template server.\n";
 	exit 1;
 }
 
-# create a default partition on the specified device on the target system
-`$LKDIR/bin/lcdremexec -d $targetSys -- "parted /dev/xvdca --script mklabel gpt mkpart xfspart xfs 0% 100%"`;
-if ($? != 0) {
-	print "Failed to create partition for $opt_r on the template server.\n";
+my $lvret = 0;
+# Check if the logical volume is created locally 
+if( system("lvdisplay '$VGName'/'$LVName' > /dev/null 2>&1") ){
+	print "Logical volume $LVName does not exist, creating now.\n";
+	
+	# Create the logical volume
+	$lvret = system("lvcreate -y --name $LVName -l 100%FREE $VGName 2>/dev/null");
+}else{
+	print "Logical volume $LVName exists, skipping create of the logical volume.\n";
+}
+
+if ($lvret != 0){
+	print "Failed to create the Logical Volume for $LVName with $VGName on ${opt_l} on the template server.\n";
 	exit 1;
 }
 
-# Make sure the file system driver module is loaded on the template
 @output = `modprobe $fsType >/dev/null 2>&1`;
 
-# Create the mirror resource on the template system (the source)
-system "$LKDRBIN/create -t $LKDRTAG -s $SWITCHBACK -p ${opt_l}1 -h \"$NETRAIDTYPE\" -n $mountPoint -x \"\" -e \"\" -f $mountPoint -y $fsType -a $syncType -b $BITMAP -z \"no\" -w \"\"";
+# Build the name for the device to mirror
+my $device = "/dev/mapper/$VGName-$LVName";
 
-$ret = $? >> 8;
+if ( ! -e $device ){
+        print "The device $device does not exist on local system, cannot proceed with mirror creation.\n";
+        exit 1;
+}
+
+#if we have resource instances for the mirror, skip trying to create the mirror
+my $inslist_ret = system("$LKBIN/ins_list -R $mountPoint > /dev/null 2>&1");
+if($inslist_ret){
+        print "Resource instance for filesystem resource with tag \"$mountPoint\" does not exist, creating resource\n";
+        $ret = system "$LKBIN/lkcli resource create dk --tag $LKDRTAG --mode $syncType --device $device --fstype $fsType --mount_point $mountPoint --fstag $mountPoint --hierarchy new";
+}else{
+        print "Resource instance for filesystem resource with tag \"$mountPoint\" already exists\n";
+        $ret = 0;
+}
+
+# Determine if mirror creation succeeded
 if ($ret != 0) {
-	print "Failed to create the scsi netraid resource hierarchy\n";
+        print "Failed to create the scsi netraid resource hierarchy\n";
```

Comment:
I would update this terminology from "scsi netraid" to "DataKeeper", because that's more obvious for a customer what that means if they are the ones looking at it

Reply:
Updated